include/RAJA/exec-openmp/target_reduce_openmp.hxx

```diff
@@ -0,0 +1,498 @@
+  ReduceMax<omp_target_reduce, T>();
+
+  const my_type * parent;
+
+  mutable T val;
+};
+
+/*!
+ ******************************************************************************
+ *
+ * \brief  Max-loc reducer class template for use in OpenMP execution.
+ *
+ *         For usage example, see reducers.hxx.
+ *
+ ******************************************************************************
+ */
+template <typename T>
+class ReduceMaxLoc<omp_target_reduce, T>
+{
+  using my_type = ReduceMaxLoc<omp_target_reduce, T>;
+
+public:
+  //
+  // Constructor takes default value (default ctor is disabled).
+  //
+  explicit ReduceMaxLoc(T init_val, Index_type init_loc):
+    parent(NULL), val(init_val), idx(init_loc)
+  {
+  }
+
+  //
+  // Copy ctor.
+  //
+  ReduceMaxLoc(const ReduceMaxLoc<omp_target_reduce, T>& other):
+    parent(other.parent ? other.parent : &other),
+    val(other.val),
+    idx(other.idx)
+  {
+  }
+
+  //
+  // Destruction releases the shared memory block chunk for reduction id
+  // and id itself for others to use.
+  //
+  ~ReduceMaxLoc<omp_target_reduce, T>()
+  {
+    if (parent) {
+#pragma omp critical
+      {
+        parent->maxloc(val, idx);
+      }
+    }
+  }
+
+  //
+  // Operator that returns reduced max value.
+  //
+  operator T()
+  {
+    return val;
+  }
+
+  //
+  // Method that returns reduced max value.
+  //
+  T get() { return operator T(); }
+
+  //
+  // Method that returns index corresponding to reduced max value.
+  //
+  Index_type getLoc()
+  {
+    return idx;
+  }
+
+  //
+  // Method that updates max and index values for current thread.
+  //
+  const ReduceMaxLoc<omp_target_reduce, T>& maxloc(T rhs, Index_type rhs_idx) const
+  {
+    if (rhs > val) {
+      val = rhs;
+      idx = rhs_idx;
+    }
+    return *this;
+  }
+
+  ReduceMaxLoc<omp_target_reduce, T>& maxloc(T rhs, Index_type rhs_idx)
+  {
+    if (rhs > val) {
+      val = rhs;
+      idx = rhs_idx;
+    }
+    return *this;
+  }
+
+private:
+  //
+  // Default ctor is declared private and not implemented.
+  //
+  ReduceMaxLoc<omp_target_reduce, T>();
+
+  const my_type * parent;
+
+  mutable T val;
+  mutable Index_type idx;
+};
+
+/*!
+ ******************************************************************************
+ *
+ * \brief  Sum reducer class template for use in OpenMP execution.
+ *
+ *         For usage example, see reducers.hxx.
+ *
+ ******************************************************************************
+ */
+
+template <typename T>
+class ReduceSum<omp_target_reduce, T>
+{
+  using my_type = ReduceSum<omp_target_reduce, T>;
+
+public:
+  //
+  // Constructor takes default value (default ctor is disabled).
+  //
+  explicit ReduceSum(T init_val, T initializer = 0)
+    : parent(NULL), val(init_val), custom_init(initializer)
+  {
+	hostid = omp_get_initial_device();
+	devid  = omp_get_default_device();
+	dev_val = new T;
+	is_mapped = false;
+	my_lev = 0;
+        dev_val = (T *)omp_target_alloc( sizeof(T), devid );
```

Comment:
I'm guessing you mean dev_val = new T not to be there, but one of them needs to go.  For this to get picked up correctly, it might be better to leave the dev_val = new T part and use `#pragma omp target enter data to(dev_val[0:1])` on it, that way it should be found correctly in the region which the omp_target_alloc value may not be.